### DISCUSSION TITLE
Working with DTO objects which already have a builder pattern

### DISCUSSION THREAD

**Discussion by Varada** [AutoFixture/AutoFixture#1230] [Q&A]
My DTO objects already use a builder pattern. Given a Dto, there is a DtoBuilder object which holds a incomplete dto object and builds out the object on .Build(). The Dto shares internal access to the setter, and only public getters. The Dto has a public constructor that can deserialize itself from a base64 encoded string representation.

Now when I try to test my methods which needs - autofixture generated DTOs, I am unable to have Autofixture generate the Dto. Autofixture tries to use the string constructor (and guid is not a valid string representation of my Dto object).

How do I nudge autofixture to treat the Dto object as just a Dto object, and use reflection to set properties that are internal? 

I did read through the issue: https://github.com/AutoFixture/AutoFixture/issues/660, but I am not able to find a way to move forward.

**Comment by Ergamon** [comment]
Is there anything to identify these Dto?

E.g. do they all share a specific naming convention like a suffix or prefix? Are they all in a specific namespace?

**Comment by Varada** [comment]
they all have a builder class with Builder suffix.

**Comment by Ergamon** [comment]
Hopefully I understood you right, your objects are basically all looking like this:

```
  public class ADto
  {
    public ADto(string serialized)
    {
      Serialized = serialized;
    }

    public string Serialized { get; }

    public string Text { get; internal set; }

    public int Number { get; internal set; }
  }
```
So if that is the case and you can identify them in any way, then you can write your own specimen builder:
```
  public class DtoSpecimenBuilder : ISpecimenBuilder
  {
    object ISpecimenBuilder.Create(object request, ISpecimenContext context)
    {
      if (request is Type type)
      {
        return Create(type, context);
      }

      return new NoSpecimen();
    }

    private object Create(Type type, ISpecimenContext context)
    {
      if (IsDto(type))
      {
        return CreateDto(type, context);
      }

      return new NoSpecimen();
    }

    private object CreateDto(Type type, ISpecimenContext context)
    {
      var dto = Activator.CreateInstance(type, "Generated for testing purposes");

      foreach (var property in type.GetProperties())
      {
        var setter = property.GetSetMethod(nonPublic: true);

        if (setter == null)
        {
          continue;
        }

        var parameters = new[]
        {
          context.Resolve(property.PropertyType),
        };

        setter.Invoke(dto, parameters);
      }

      return dto;
    }


    private static bool IsDto(Type type)
    {
      return type.Name.EndsWith("Dto");
    }
  }
```
(Your IsDto method must probably be different, looking for this builder)
But with this in hand you can setup your fixtures like this:

```
      var fixture = new Fixture();
      fixture.Customizations.Add(new DtoSpecimenBuilder());
      var dto = fixture.Create<ADto>();
```
Just an idea from what I have read
